apps/etterlatte-beregning/src/main/kotlin/beregning/grunnlag/BarnepensjonsBeregningGrunnlagMigreringsJob.kt

```diff
@@ -0,0 +1,127 @@
+    private val dataSource: DataSource,
+    private val featureToggleService: FeatureToggleService
+) {
+    private val log = LoggerFactory.getLogger(this::class.java)
+
+    private val config: Config = ConfigFactory.load()
+
+    private val klient = httpClientClientCredentials(
+        azureAppClientId = config.getString("AZURE_APP_CLIENT_ID"),
+        azureAppJwk = config.getString("AZURE_APP_JWK"),
+        azureAppWellKnownUrl = config.getString("AZURE_APP_WELL_KNOWN_URL"),
+        azureAppScope = config.getString("BEREGNING_AZURE_SCOPE")
+    )
+
+    private val resourceUrl = config.getString("grunnlag.resource.url")
+
+    fun run() {
+        if (!featureToggleService.isEnabled(MigreringsJobFeatureToggle.BeregningsGrunnlagFraBeregning, false)) {
+            return
+        }
+
+        val correlationId = UUID.randomUUID().toString()
+
+        if (leaderElection.isLeader()) {
+            withLogContext(correlationId) { log.info("Starter jobb ${this::class.simpleName}") }
+
+            using(sessionOf(dataSource)) { session ->
+                session.transaction { tx ->
+
+                    if (trengsMigrering(tx)) {
+                        withLogContext(correlationId) {
+                            log.info("Migrering trengs")
+
+                            hentAlleBeregninger(tx).forEach { runBlocking { process(correlationId, it) } }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun hentAlleBeregninger(tx: TransactionalSession) = tx.run(
+        queryOf("SELECT behandlingId, grunnlagVersjon, sakId FROM beregningsperiode").map {
+            BeregningForMigrering(
+                it.uuid("behandlingId"),
+                it.long("grunnlagVersjon"),
+                it.long("sakId")
+            )
+        }.asList
+    )
+
+    private fun trengsMigrering(tx: TransactionalSession) = (
+        tx.run(
+            queryOf("SELECT COUNT(*) FROM bp_beregningsgrunnlag").map { it.long(0) }.asSingle
+        ) ?: 0
+        ) == 0L
+
+    private suspend fun process(correlationId: String, beregning: BeregningForMigrering) {
+        val response = klient.get(
+            "$resourceUrl/api/grunnlag/${beregning.sakId}/versjon/${beregning.grunnlagVersjon}"
+        ) {
+            header(X_CORRELATION_ID, correlationId)
+        }
+
+        if (response.status.isSuccess()) {
+            val grunnlag = response.body<Grunnlag>().mapSoeskenJustering()
+
+            grunnlag?.let {
+                beregingsGrunnlagRepository.lagre(
+                    BeregningsGrunnlag(
+                        beregning.behandlingId,
+                        it.kilde as Grunnlagsopplysning.Saksbehandler,
+                        it.verdi.beregningsgrunnlag
+                    )
+                )
+            }
+        } else {
+            throw ResponseException(response, "Ukjent feil fra grunnlag api")
+        }
+    }
+}
```

Comment:
Denne metoden bør logge veldig eksplisitt at den har migrert for behandlingen med id i loggen, så er det lettere for oss å plukke opp feil. Den burde også logge eksplisitt hvis den feiler, og ikke bare boble opp en exception. 

Så vidt jeg forstår flyten i koden her er det tenkt å kjøre all migreringen i en transaction, slik at man får migrert alle eller ingen -- stemmer dette? Jeg tror det er en rimelig måte å få gjort det på, så vi unngår å havne i en delvis migrert virkelighet. Savner bare litt logging for å følge med på at det går som det skal :)

Reply:
Skal legge til logging.

Målet er en transaction ja. Men - sånn det er satt opp her så får vi ikke det helt til.

Hvis jeg kopiere koden i repository ut og gjør det her med samme transaction så vil det fungere.

Men - vi har ikke løst "gjenbruk av eksisterende transaction" enda.

Reply:
Flytter koden inn for gjenbruk av tx